google/cloud/spanner/row.h

```diff
@@ -16,14 +16,186 @@
 #define GOOGLE_CLOUD_CPP_SPANNER_GOOGLE_CLOUD_SPANNER_ROW_H_
 
 #include "google/cloud/spanner/internal/tuple_utils.h"
+#include "google/cloud/spanner/value.h"
 #include "google/cloud/spanner/version.h"
+#include "google/cloud/status.h"
+#include "google/cloud/status_or.h"
+#include <memory>
+#include <string>
+#include <utility>
+#include <vector>
 
 namespace google {
 namespace cloud {
 namespace spanner {
 inline namespace SPANNER_CLIENT_NS {
 
-// TODO(#387)
+/**
+ * A `Row` is a sequence of columns each with a name and an associated `Value`.
+ *
+ * The `Row` class is a regular value type that may be copied, moved, assigned,
+ * compared for equality, etc. Instances may be large if they hold lots of
+ * `Value` data, so copy only when necessary.
+ *
+ * `Row` instances are typically returned as the result of queries or reads of
+ * a Cloud Spanner table (see `Client::Read` and `Client::ExecuteQuery`). Users
+ * will mostly just use the accessor methods on `Row, and will rarely (if ever)
+ * need to construct a `Row` of their own.
+ *
+ * The number of columns in a `Row` can be obtained from the `size()` member
+ * function. The `Value`s can be obtained using the `values()` accessor. The
+ * names of each column in the row can be obtained using the `columns()`
+ * accessor.
+ *
+ * Perhaps the most convenient way to access the `Values` in a row is through
+ * the variety of "get" accessors. A user may access a column's `Value' by
+ * calling `get` with a `std::size_t` 0-indexed position, or a `std::string`
+ * column name. Furthermore, callers may directly extract the native C++ type
+ * by specifying the C++ type along with the column's position or name.
+ *
+ * @par Example
+ *
+ * @code
+ * Row row = ...;
+ * if (StatusOr<std::string> x = row.get<std::string>("LastName")) {
+ *   std::cout << "LastName=" << *x << "\n";
+ * }
+ * @endcode
+ *
+ * @note There is a helper function defined below named `MakeRow()` to make
+ *     creating `Row` instances for testing easier.
+ */
+class Row {
+ public:
+  /// Default constructs an empty row with no columns nor values.
+  Row();
+
+  /**
+   * Constructs a `Row` with the given @p values and @p columns.
+   *
+   * @note columns must not be nullptr
+   * @note columns.size() must equal values.size()
+   */
+  explicit Row(std::vector<Value> values,
+               std::shared_ptr<const std::vector<std::string>> columns);
```

Comment:
I spent some time today to think about how we might want to adapt this for the BigQuery client. I'm replying here so you're aware of the current thinking and if there are problems with our approach, you can let us know.

BigQuery returns data in response messages containing one or more rows in serialized Arrow format. We have a desire to not fully materialize every row as that can be quite inefficient. What I'm thinking about is an internal-only constructor like:

```cpp
Row(
  std::shared_ptr<arrow::Buffer> buf /* contains one of more row */,
  std::shared_ptr<arrow::Schema> schema,
  std::size_t offset_in_buf /* i.e., this is the row # in buf */ );
```

This construction defers all parsing until the user wants to fetch a column value (and even then, the parsing that is done is quite efficient).

One problem with this approach that we *may* need to solve later: if the user is reading the data, but sampling rows into a container, then the memory consumption will not be equivalent to the number of rows sampled, but equivalent to the number of arrow:Buffer objects that have to be kept around for those sampled rows. If this becomes a problem, we can add a function to materialize a row such that it doesn't rely on the shared buffer.

Another problem we'll need to figure out is how to map non-primitive types (e.g., lists, unions, structs) into C++ types without materializing too much data. For example, we should be able to have an STL-friendly list that is lazy. I noticed that the Value type in Spanner uses a vector, which is probably not great for us, though we can start with such an implementation to start. Let me know if you have any thoughts on this.

Thanks!